proposed/0023-crust-mock/0023-crust-mock.md

```diff
@@ -0,0 +1,106 @@
+- Feature Name: Crust mock
+- Type: feature
+- Related components: crust, routing
+- Start Date: 18-02-2016
+- RFC PR:
+- Issue number:
+
+# Summary
+
+Drop-in mock replacement for Crust to make it easier to unit test libraries and
+applications that use Crust.
+
+# Motivation
+
+Testing Crust-depending libraries on real network (even on LAN or localhost) is
+cumbersome, slow and potentially expensive. A drop in mock replacement for Crust
+could make these test simpler and faster, make it easier to diagnose failures,
+and help cover edge cases difficult to set up on real network.
+
+# Detailed design
+
+This section details one possible implementation of the Crust Mock.
+
+## Changes to crust
+
+A new trait `ConnectionService` MUST be defined. It MUST have all the relevant methods of `crust::Service`:
+
+```rust
+pub trait ConnectionService {
+  fn start_listening_tcp();
+  fn start_listening_utp();
+  fn start_service_discovery();
+  fn prepare_connection_info();
+  fn connect();
+  fn disconnect();
+  fn send();
+}
+```
+
+This trait MUST be implemented by the current `crust::Service` and by the new
+`crust::mock::Service`.
+
+## Changes to routing
+
+`routing::Core` MUST be either made generic over the `ConnectionService`, or change the type of its `crust_service` member to `Box<ConnectionService>`.
+
+A way to specify which instance of `ConnectionService` to pass into `routing::Core` MUST be devised. One possible option is to have `routing::Core::new` accept a `Fn` which returns the `ConnectionService` instance when called with `CrustEventSender`:
+
+```rust
+routing::Core::new(..., |crust_event_sender| {
+  Box::new(crust::mock::Service::new(crust_event_sender))
+})
+```
+
+## Features of `crust::mock::Service`
+
+- Do not hit real network (not even LAN or localhost), do not use TCP/UDP sockets.
+- Simulate all the relevant config files crust uses (crust config, bootstrap cache, etc...). Possibly by accepting structs, representing those configs, as
+parameters to the constructor.
```

Comment:
If all three are `<Option>` Then it can satisfy this and make testing much simpler. Also allows users to not use the config_file_handler if they do not wish. For our apps it's easier to have a common paradigm like config_file_handler etc. so best of both worlds perhaps?